airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbRecordConsumer.java

```diff
@@ -0,0 +1,164 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.MongoDatabase;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.base.AirbyteMessageConsumer;
+import io.airbyte.integrations.base.AirbyteStreamNameNamespacePair;
+import io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.bson.Document;
+import org.joda.time.LocalDateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MongodbRecordConsumer extends FailureTrackingAirbyteMessageConsumer implements AirbyteMessageConsumer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongodbRecordConsumer.class);
+
+  private static final String AIRBYTE_DATA = "_airbyte_data";
+  private static final String AIRBYTE_EMITTED_AT = "_airbyte_emitted_at";
+
+  private final Map<AirbyteStreamNameNamespacePair, MongodbWriteConfig> writeConfigs;
+  private final MongoDatabase mongoDatabase;
+  private final ConfiguredAirbyteCatalog catalog;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
+  private final ObjectMapper objectMapper;
+
+  private AirbyteMessage lastStateMessage = null;
+
+  public MongodbRecordConsumer(Map<AirbyteStreamNameNamespacePair, MongodbWriteConfig> writeConfigs,
+                               MongoDatabase mongoDatabase,
+                               ConfiguredAirbyteCatalog catalog,
+                               Consumer<AirbyteMessage> outputRecordCollector) {
+    this.writeConfigs = writeConfigs;
+    this.mongoDatabase = mongoDatabase;
+    this.catalog = catalog;
+    this.outputRecordCollector = outputRecordCollector;
+    this.objectMapper = new ObjectMapper();
+  }
+
+  @Override
+  protected void startTracked() {
+    // todo (cgardens) - move contents of #write into this method.
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage message) {
+    if (message.getType() == AirbyteMessage.Type.STATE) {
+      lastStateMessage = message;
+    } else if (message.getType() == AirbyteMessage.Type.RECORD) {
+      final AirbyteRecordMessage recordMessage = message.getRecord();
+      AirbyteStreamNameNamespacePair pair = AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage);
+
+      if (!writeConfigs.containsKey(pair)) {
+        throw new IllegalArgumentException(
+            String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
+                Jsons.serialize(catalog), Jsons.serialize(recordMessage)));
+      }
+      try {
```

Comment:
You can extract try-catch block to a separate method for better readability.

Reply:
Yes, the code looks much better this way.